test/README.md

```diff
@@ -12,9 +14,219 @@ To run all unit tests:
 go test ./...
 ```
 
+## Test library
+
+You can use the test library in this dir to:
+
+* [Use common test flags](#use-common-test-flags)
+* [Output logs](#output-logs)
+* [Emit metrics](#emit-metrics)
+* [Ensure test cleanup](#ensure-test-cleanup)
+
+### Use common test flags
+
+These flags are useful for running against an existing cluster, making use of your existing
+[environment setup](https://github.com/knative/serving/blob/master/DEVELOPMENT.md#environment-setup).
+
+By importing `github.com/knative/serving/test` you get access to a global variable called
+`test.Flags` which holds the values of [the command line flags](/test/README.md#flags).
+
+```go
+imagePath := strings.Join([]string{test.ServingFlags.DockerRepo, image}, "/"))
+```
+
+_See [e2e_flags.go](./e2e_flags.go)._
+
+
+### Output logs
+
+[When tests are run with `--logverbose`
+option](README.md#output-verbose-logs), debug logs will be emitted to stdout.
+
+We are using the common [e2e logging library](logging/logging.go) that uses the [Knative logging library](../logging/) for structured logging. 
+It is built on top of [zap](https://github.com/uber-go/zap). Tests should initialize the global logger to use a test specifc context with `logging.GetContextLogger`:
+
+```go
+// The convention is for the name of the logger to match the name of the test.
+logging.GetContextLogger("TestHelloWorld")
+```
+
+Logs can then be emitted using the `logger` object which is required by
+many functions in the test library. To emit logs:
+
+```go
+logger.Infof("Creating a new Route %s and Configuration %s", route, configuration)
+logger.Debugf("The LogURL is %s, not yet verifying", logURL)
+```
+
+_See [logging.go](./logging/logging.go)._
+
+### Emit metrics
+
+You can emit metrics from your tests using [the opencensus
+library](https://github.com/census-instrumentation/opencensus-go), which [is being
+used inside Knative as well](https://github.com/knative/serving/blob/master/docs/telemetry.md).
+These metrics will be emitted by the test if the test is run with [the `--emitmetrics` option](#metrics-flag).
+
+You can record arbitrary metrics with
+[`stats.Record`](https://github.com/census-instrumentation/opencensus-go#stats) or
+measure latency with [`trace.StartSpan`](https://github.com/census-instrumentation/opencensus-go#traces):
+
+```go
+ctx, span := trace.StartSpan(context.Background(), "MyMetric")
+```
+
+* These traces will be emitted automatically by [the generic crd polling
+  functions](#check-knative-serving-resources).
+* The traces are emitted by [a custom metric exporter](./logging/logging.go)
+  that uses the global logger instance.
+
+#### Metric format
+
+When a `trace` metric is emitted, the format is `metric name startTime endTime duration`. The name
+of the metric is arbitrary and can be any string. The values are:
+
+* "metric" - Indicates this log is a metric
```

Comment:
remove quotes?

Reply:
hm i think this was trying to convent that it is literally the word `metric`, while the others are placeholders. I'll try to make this more clear!